posts/08-ic-xnet.html.pm

```diff
@@ -0,0 +1,237 @@
+}
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-subnets"]{
+    Two three-node subnets located in three data centers.
+    Solid lines represent peer-to-peer communication within a subnet; dotted lines — cross-subnet communication.
+    Not all XNet connections are present in the picture.
+  }
+  ◊p{◊(embed-svg "images/08-subnets.svg")}
+}
+◊p{
+  XNet protocol is not limited to the same-datacenter communication; any node from one subnet can talk to any other node from another subnet.
+  However, the replica prefers contacting close◊sidenote["mn-proximity"]{
+    The replica uses network latency as the measure of proximity.
+    It randomly contacts nodes from other subnets , assigning higher weights to nodes with lower latency.
+  } peers to reduce the message delivery latency.
+}
+}
+
+◊section{
+◊section-title["block-payloads"]{Block payloads}
+◊p{
+  The job of the consensus protocol is to aggregate messages from the outside world and pack them into a neat block.
+  The consensus includes several types of messages into blocks, such as user ingress messages, Bitcoin transactions (for subnets with enabled Bitcoin integration), and inter-canister messages from other subnets.
+  We call messages paired with the data required for their validation a ◊em{payload}.
+  We call components that pull payloads from the network ◊em{payload builders}.
+}
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-payloads"]{The consensus algorithm aggregates messages from the outside world into blocks.}
+  ◊p{◊(embed-svg "images/08-block-payloads.svg")}
+}
+◊p{
+  XNet payload builder pulls messages from nodes assigned to other subnets using a simple HTTP protocol.
+  ◊em{XNet endpoint} is a component that serves messages destined for other subnets over secure TLS connections, accepting connections only from other nodes◊sidenote["sn-xnet-tls"]{This measure does not imply privacy because malicious nodes can access the data; but ensures that network providers cannot read the messages.}.
+  XNet endpoint fetches the complete list of nodes, their subnet assignment, IP addresses, and public keys (required to establish a TLS connection) from the registry.
+}
+
+◊figure[#:class "grayscale-diagram"]{
+  ◊marginnote["mn-endpoint"]{
+    The data flow in the XNet protocol.
+    Subnet ◊math{Y} produces a signed stream of messages for subnet X and exposes this stream via an HTTP endpoint.
+    Subnet ◊math{X} pulls messages from one of the nodes on subnet Y.
+  }
+  ◊p{◊(embed-svg "images/08-xnet-endpoint.svg")}
+}
+}
+
+◊section{
+◊section-title["message-streams"]{Message streams}
+◊p{
+  Messages start their journey in the output queues of canisters hosted on the source subnet.
+  The subnet sorts these messages based on the destination subnet and conflates them into a flat message stream (one stream per destination subnet).
```

Comment:
"Conflate" has a negative connotation in my head. Wikipedia seems to agree. I'd go with "merges" or "routes".

Also "a flat message stream" -> "flat message streams".

Reply:
Good point. I replaced "conflates" with "interleaves".